metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -548,6 +550,15 @@ static void validateNewTopicNames(Map<String, ApiError> topicErrors,
                 topicErrors.put(topic.name(),
                     new ApiError(Errors.INVALID_TOPIC_EXCEPTION, e.getMessage()));
             }
+            if (Topic.hasCollisionChars(topic.name())) {
```

Comment:
We shouldn't loop through all topics in the cluster here. This could be a very long list. Instead, let's keep a separate "potentially colliding names" set of strings where all the colliding characters are translated to a single character.